### DISCUSSION TITLE
Installation via Anaconda fails

### DISCUSSION THREAD

**Discussion by iainhallam** [mesa/mesa#1774] [Q&A]
I've got a fresh Anaconda 3 installation (2023.07-2) and I get some syntax errors when trying to install Mesa into it. I'm a Linux admin, rather than Python programmer, so can anyone see what's likely to be the cause, please? I'm on Rocky 8 (clone of Red Hat EL8) and have installed the Anaconda prerequisites using yum.

```bash
$ rm -rf /opt/anaconda/3-2023  # Clears all previous installations; I have write permissions on /opt/anaconda
$ bash Downloads/Anaconda3-2023.07-2-Linux-x86_64.sh -b -f -p /opt/anaconda/3-2023
<...>
installation finished.
$ /opt/anaconda/3-2023/bin/activate
$ /opt/anaconda/3-2023/bin/conda install --verbose --channel conda-forge --yes mesa
<...>
The following NEW packages will be INSTALLED:

  mesa               conda-forge/noarch::mesa-2.1.1-pyhd8ed1ab_0 
  python_abi         conda-forge/linux-64::python_abi-3.11-2_cp311 

The following packages will be UPDATED:

  ca-certificates    pkgs/main::ca-certificates-2023.05.30~ --> conda-forge::ca-certificates-2023.7.22-hbcca054_0 
  conda              pkgs/main::conda-23.7.2-py311h06a4308~ --> conda-forge::conda-23.7.3-py311h38be061_0 
  openssl                                 1.1.1u-h7f8727e_0 --> 1.1.1v-h7f8727e_0 

The following packages will be SUPERSEDED by a higher-priority channel:

  certifi            pkgs/main/linux-64::certifi-2023.7.22~ --> conda-forge/noarch::certifi-2023.7.22-pyhd8ed1ab_0
<...>
INFO conda.gateways.disk.create:compile_multiple_pyc(455): pyc file failed to compile successfully (run_command failed)
python_exe_full_path: /opt/anaconda/3-2023/bin/python3.11
py_full_path: /opt/anaconda/3-2023/lib/python3.11/site-packages/mesa/visualization/modules/BarChartVisualization.py
pyc_full_path: /opt/anaconda/3-2023/lib/python3.11/site-packages/mesa/cookiecutter-mesa/{{cookiecutter.snake}}/{{cookiecutter.snake}}/__pycache__/server.cpython-311.pyc
compile rc: 1
compile stdout: *** Error compiling 'lib/python3.11/site-packages/mesa/cookiecutter-mesa/{{cookiecutter.snake}}/run.py'...
  File "lib/python3.11/site-packages/mesa/cookiecutter-mesa/{{cookiecutter.snake}}/run.py", line 1
    from {{cookiecutter.snake}}.server import server  # noqa
         ^
SyntaxError: invalid syntax
```

There are another 8 of these errors, all at `{{cookiecutter.<something>}}` strings.

Do let me know if this is the wrong place to ask about this, too!

**Comment by rht** [comment]
Do you have to use `conda`? For now, can you use `pip`? What if you use Docker inside of Rocky Linux? https://github.com/projectmesa/mesa/pull/1757 for how to do it.

This seems like a bug instead of discussion. The reason is somehow `conda` doesn't like filenames with `{{...}}`.

**Reply by rht** [reply]
Oh, it's not the filename, it's the template inside the cookiecutter files.

**Reply by rht** [reply]
https://github.com/conda/conda-build/issues/686 has a solution for it,

**Reply by iainhallam** [reply]
Reading through that issue and the linked docs, it seems like the `skip_compile_pyc` option would need to be added to the Mesa Anaconda package recipe. Is that maintained by Project Mesa, or someone else? I.e., where can I open an issue for that?

Thanks for looking at this - I manage 300 Linux PCs for a university, and Anaconda is used by lots of courses for their Python environment, so getting to install using conda would be much better than other options. Docker's not available on these shared-access machines.

**Reply by rht** [reply]
@jackiekazil would know better.

For now, what about https://stackoverflow.com/questions/19042389/conda-installing-upgrading-directly-from-github ?
You can specify a dependency that is installed via `pip` with `conda-env`. Or just run `pip install` inside the conda env: https://stackoverflow.com/questions/41060382/using-pip-to-install-packages-to-anaconda-environment.

**Reply by jackiekazil** [reply]
I usually pip, but maybe this will be helpful -- https://github.com/conda-forge/mesa-feedstock

**Reply by rht** [reply]
This feels like a general problem that affects any package that ships with cookiecutter templates. Maybe there is a standardized solution that is different from `skip_compile_pyc`. Currently digging into this, to make sure there is no stone left unturned.

**Reply by rht** [reply]
There is https://github.com/cookiecutter/cookiecutter/issues/1049. I had thought about renaming the .py files to something else, since technically they are not actual Python files with correct syntax anyway. The issue OP seemed to do this as a workaround. But I think it is a more proper solution.

**Reply by rht** [reply]
Should be fixed in #1785. @jackiekazil @tpike3 should we release 2.1.2 very soon, since it has accumulated enough fixes?

**Reply by tpike3** [reply]
Merged and agree we should do another release.

**Reply by iainhallam** [reply]
That's great news, @rht, thanks. Do you know what sort of timetable a new release would be?

**Reply by rht** [reply]
Hopefully by the end of this month: https://github.com/projectmesa/mesa/discussions/1787#discussioncomment-6955546. Only 2 issues left:
- #1768 is not critical and could be skipped
- #1792 is a blocker, but is a quick fix. Only few lines of code of change is needed.

**Comment by jackiekazil** [comment]
Agree

**Comment by rht** [comment]
@iainhallam we just released 2.1.2 with the cookiecutter fix. Can you try this again?
